File: packages/tolk/test/parity/elementwise_2d/main.py
#!/usr/bin/env python3
"""Parity case: c[i*16+j] = a[i*16+j] + b[i*16+j], 2 Global ranges. GPU-only."""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from helpers import dump, GPU_BACKENDS  # noqa: E402

from tinygrad.uop.ops import UOp, Ops, KernelInfo, AxisType  # noqa: E402
from tinygrad.dtype import dtypes  # noqa: E402


def kernel():
    ROWS, COLS = 8, 16
    p0 = UOp.param(0, dtypes.float32, shape=(-1,))
    p1 = UOp.param(1, dtypes.float32, shape=(-1,))
    p2 = UOp.param(2, dtypes.float32, shape=(-1,))
    ri = UOp.range(ROWS, 0, AxisType.GLOBAL)
    rj = UOp.range(COLS, 1, AxisType.GLOBAL)
    flat = ri * COLS + rj
    ld_a = p0.index(flat).load()
    ld_b = p1.index(flat).load()
    add = ld_a + ld_b
    st = p2.index(flat).store(add)
    end = st.end(ri, rj)
    return UOp.sink(
        end,
        arg=KernelInfo(
            name="elementwise_2d",
            axis_types=(AxisType.GLOBAL, AxisType.GLOBAL),
            opts_to_apply=(),
        ),
    )


if __name__ == "__main__":
    dump(kernel(), os.path.dirname(os.path.abspath(__file__)),
         stages=("stage5", "stage7"), backends=GPU_BACKENDS)

File: packages/tolk/test/parity/elementwise_2d/main.ml
(* Parity case: c[i*16+j] = a[i*16+j] + b[i*16+j], 2 Global ranges. GPU-only. *)

open Tolk_uop
module U = Uop

let kernel () =
  let rows, cols = (8, 16) in
  let p0 = U.param ~slot:0 ~dtype:Helpers.global_fptr () in
  let p1 = U.param ~slot:1 ~dtype:Helpers.global_fptr () in
  let p2 = U.param ~slot:2 ~dtype:Helpers.global_fptr () in
  let ri = U.range ~size:(Helpers.idx rows) ~axis:0 ~kind:Axis_type.Global () in
  let rj = U.range ~size:(Helpers.idx cols) ~axis:1 ~kind:Axis_type.Global () in
  let open U.O in
  let flat = (ri * int_ cols) + rj in
  let ld_a = U.load ~src:(U.index ~ptr:p0 ~idxs:[flat] ()) () in
  let ld_b = U.load ~src:(U.index ~ptr:p1 ~idxs:[flat] ()) () in
  let add = U.alu_binary ~op:Ops.Add ~lhs:ld_a ~rhs:ld_b in
  let st =
    U.store ~dst:(U.index ~ptr:p2 ~idxs:[flat] ()) ~value:add ()
  in
  let e = U.end_ ~value:st ~ranges:[ ri; rj ] in
  U.sink
    ~kernel_info:
      {
        U.name = "elementwise_2d";
        axis_types = [ Axis_type.Global; Axis_type.Global ];
        dont_use_locals = false;
        applied_opts = [];
        opts_to_apply = Some [];
        estimates = None;
      beam = 0;
      }
    [ e ]

let () =
  Helpers.dump ~backends:Helpers.gpu_backends
    ~stages:[ Helpers.Stage5; Helpers.Stage7 ]
    ~out_dir:Sys.argv.(1) (kernel ())
